In [1]:
# import the necessary libraries
import pandas as pd
import numpy as np
from datetime import datetime

In [2]:
#Define the path
path = "C:/Users/lunam/OzonProject/"

# Cleaning Data

### Cleaning Weather Data 

In [3]:
# read the text file into a dataframe
wetter = pd.read_csv(path + "order_116936_data.txt", sep=";", header=None)

In [4]:
#Let's start by analysing our dataset's structure
wetter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76611 entries, 0 to 76610
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       76611 non-null  object
dtypes: object(1)
memory usage: 598.7+ KB


In [5]:
#Let's take a look at the dataset itself
wetter.head()

0
0  stn;time;prestah0;tre200h0;fu3010h0
1       ZHROS;2021010100;954.3;2.1;1.1
2       ZHROS;2021010101;954.3;1.9;1.1
3       ZHROS;2021010102;954.7;2.2;1.4
4       ZHROS;2021010103;954.9;2.0;1.1

Aufgrund der Natur des Datenfiles müssen wir erstens das Format anpassen.

In [6]:
#We only take the station STA, so we need to cut off the firs 51073 rows of the dataset
wetter = wetter.iloc[51074:]

In [7]:
wetter

0
51074  stn;time;gre000h0;prestah0;tre200h0;fu3010h0
51075              ZHSTA;2021010100;0;953.9;2.4;4.3
51076              ZHSTA;2021010101;0;954.0;2.1;4.7
51077              ZHSTA;2021010102;0;954.4;2.3;3.2
51078              ZHSTA;2021010103;0;954.6;2.0;2.5
...                                             ...
76606              ZHSTA;2023113019;0;948.9;3.2;2.9
76607              ZHSTA;2023113020;0;948.9;3.8;5.4
76608              ZHSTA;2023113021;0;949.1;4.3;6.8
76609              ZHSTA;2023113022;0;949.2;3.8;5.4
76610              ZHSTA;2023113023;0;949.1;4.6;4.7

[25537 rows x 1 columns]

In [8]:
#split the single column into multiple columns using the semicolon as a separator
wetter = wetter[0].str.split(';', expand=True)
wetter

0           1         2         3         4         5
51074    stn        time  gre000h0  prestah0  tre200h0  fu3010h0
51075  ZHSTA  2021010100         0     953.9       2.4       4.3
51076  ZHSTA  2021010101         0     954.0       2.1       4.7
51077  ZHSTA  2021010102         0     954.4       2.3       3.2
51078  ZHSTA  2021010103         0     954.6       2.0       2.5
...      ...         ...       ...       ...       ...       ...
76606  ZHSTA  2023113019         0     948.9       3.2       2.9
76607  ZHSTA  2023113020         0     948.9       3.8       5.4
76608  ZHSTA  2023113021         0     949.1       4.3       6.8
76609  ZHSTA  2023113022         0     949.2       3.8       5.4
76610  ZHSTA  2023113023         0     949.1       4.6       4.7

[25537 rows x 6 columns]

In [9]:
#make the top row into the header
wetter.columns = wetter.iloc[0]
wetter = wetter[1:]
wetter

51074    stn        time gre000h0 prestah0 tre200h0 fu3010h0
51075  ZHSTA  2021010100        0    953.9      2.4      4.3
51076  ZHSTA  2021010101        0    954.0      2.1      4.7
51077  ZHSTA  2021010102        0    954.4      2.3      3.2
51078  ZHSTA  2021010103        0    954.6      2.0      2.5
51079  ZHSTA  2021010104        0    954.6      2.3      2.2
...      ...         ...      ...      ...      ...      ...
76606  ZHSTA  2023113019        0    948.9      3.2      2.9
76607  ZHSTA  2023113020        0    948.9      3.8      5.4
76608  ZHSTA  2023113021        0    949.1      4.3      6.8
76609  ZHSTA  2023113022        0    949.2      3.8      5.4
76610  ZHSTA  2023113023        0    949.1      4.6      4.7

[25536 rows x 6 columns]

In [10]:
#change the date variable to a useful format
wetter['time'] = wetter['time'].str[:-6] + '-' + wetter['time'].str[-6:]
wetter['time'] = wetter['time'].str[:-4] + '-' + wetter['time'].str[-4:]
wetter['time'] = wetter['time'].str[:-2] + ' ' + wetter['time'].str[-2:]
wetter['time'] = wetter['time'] + ':00:00'

C:\Users\lunam\AppData\Local\Temp\ipykernel_21600\3649913782.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wetter['time'] = wetter['time'].str[:-6] + '-' + wetter['time'].str[-6:]
C:\Users\lunam\AppData\Local\Temp\ipykernel_21600\3649913782.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wetter['time'] = wetter['time'].str[:-4] + '-' + wetter['time'].str[-4:]
C:\Users\lunam\AppData\Local\Temp\ipykernel_21600\3649913782.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice

In [11]:
wetter

51074    stn                 time gre000h0 prestah0 tre200h0 fu3010h0
51075  ZHSTA  2021-01-01 00:00:00        0    953.9      2.4      4.3
51076  ZHSTA  2021-01-01 01:00:00        0    954.0      2.1      4.7
51077  ZHSTA  2021-01-01 02:00:00        0    954.4      2.3      3.2
51078  ZHSTA  2021-01-01 03:00:00        0    954.6      2.0      2.5
51079  ZHSTA  2021-01-01 04:00:00        0    954.6      2.3      2.2
...      ...                  ...      ...      ...      ...      ...
76606  ZHSTA  2023-11-30 19:00:00        0    948.9      3.2      2.9
76607  ZHSTA  2023-11-30 20:00:00        0    948.9      3.8      5.4
76608  ZHSTA  2023-11-30 21:00:00        0    949.1      4.3      6.8
76609  ZHSTA  2023-11-30 22:00:00        0    949.2      3.8      5.4
76610  ZHSTA  2023-11-30 23:00:00        0    949.1      4.6      4.7

[25536 rows x 6 columns]

In [12]:
#We change the column name from time to datetime
wetter = wetter.rename(columns={"time" : "datetime"})

In [13]:
wetter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25536 entries, 51075 to 76610
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   stn       25536 non-null  object
 1   datetime  25536 non-null  object
 2   gre000h0  25536 non-null  object
 3   prestah0  25536 non-null  object
 4   tre200h0  25536 non-null  object
 5   fu3010h0  25536 non-null  object
dtypes: object(6)
memory usage: 1.2+ MB


In [14]:
#We drop the column gre000h0, as after inspecting the raw dataset we noticed that there were too many missing values
wetter.drop(columns="gre000h0", inplace=True)

In [15]:
#We rename the columns into a more understandable label
wetter = wetter.rename(columns={"prestah0": "Luftdruck_STA", "tre200h0": "Lufttemp_STA", "fu3010h0": "Windgeschw_STA"})

In [16]:
#We replace missing values with NaN
wetter["Luftdruck_STA"] = wetter["Luftdruck_STA"].replace('-', np.nan)
wetter["Lufttemp_STA"] = wetter["Lufttemp_STA"].replace('-', np.nan)
wetter["Windgeschw_STA"] = wetter["Windgeschw_STA"].replace('-', np.nan)

In [17]:
#We drop the column stn (station)
wetter = wetter.drop(columns='stn')

In [18]:
#We set datetime as index
wetter.set_index("datetime", inplace=True)

In [19]:
#We make certain columns to floats to work with them
columns_num_w = ["Luftdruck_STA", "Lufttemp_STA", "Windgeschw_STA"]
wetter[columns_num_w] = wetter[columns_num_w].apply(pd.to_numeric, errors='coerce')

In [20]:
wetter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25536 entries, 2021-01-01 00:00:00 to 2023-11-30 23:00:00
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Luftdruck_STA   25416 non-null  float64
 1   Lufttemp_STA    25287 non-null  float64
 2   Windgeschw_STA  25406 non-null  float64
dtypes: float64(3)
memory usage: 798.0+ KB


In [21]:
#We impute missing values using spline interpolate imputation, which is used for time series
all_variables_w = (list(wetter.columns))
wetter = wetter[all_variables_w].interpolate(option='spline')

In [22]:
#We get info of final variables in dataset
wetter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25536 entries, 2021-01-01 00:00:00 to 2023-11-30 23:00:00
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Luftdruck_STA   25536 non-null  float64
 1   Lufttemp_STA    25536 non-null  float64
 2   Windgeschw_STA  25536 non-null  float64
dtypes: float64(3)
memory usage: 798.0+ KB


In [23]:
wetter.head()

51074                Luftdruck_STA  Lufttemp_STA  Windgeschw_STA
datetime                                                        
2021-01-01 00:00:00          953.9           2.4             4.3
2021-01-01 01:00:00          954.0           2.1             4.7
2021-01-01 02:00:00          954.4           2.3             3.2
2021-01-01 03:00:00          954.6           2.0             2.5
2021-01-01 04:00:00          954.6           2.3             2.2

In [24]:
#we drop the first row, since as we are going to see later, we take for Ozon the end value
wetter = wetter.drop(wetter.index[0])

In [25]:
wetter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25535 entries, 2021-01-01 01:00:00 to 2023-11-30 23:00:00
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Luftdruck_STA   25535 non-null  float64
 1   Lufttemp_STA    25535 non-null  float64
 2   Windgeschw_STA  25535 non-null  float64
dtypes: float64(3)
memory usage: 798.0+ KB


In [26]:
wetter.head()

51074                Luftdruck_STA  Lufttemp_STA  Windgeschw_STA
datetime                                                        
2021-01-01 01:00:00          954.0           2.1             4.7
2021-01-01 02:00:00          954.4           2.3             3.2
2021-01-01 03:00:00          954.6           2.0             2.5
2021-01-01 04:00:00          954.6           2.3             2.2
2021-01-01 05:00:00          954.8           2.3             2.9

Jetzt ist unser Dataset mit Wetterdaten bereit und vollständig. Wir merken, dass das Dataset 25535 Spalten hat. Diese ist die Anzahl von Studen innerhalb 2 Jahren und 11 Monate (2021-01-01 01:00:00 to 2023-11-30 23:00:00).

### Cleaning Ozon Data

In [27]:
#We load the dataset
ozon = pd.read_csv(path + "Ozon.csv", sep=";")

In [28]:
#We delete the first three rows
ozon = ozon.drop(index=range(3))
#We delete the last two rows
ozon = ozon.iloc[:-2]
#We delete the "Startzeit" since we are going to use the value at the end of each hour (Endbestand)
ozon = ozon.drop(["Startzeit"], axis = 1)
#We change date into useful format
ozon["Endzeit"] = pd.to_datetime(ozon["Endzeit"], format='%d.%m.%Y %H:%M')
ozon["Endzeit"] = ozon["Endzeit"].dt.strftime('%Y-%m-%d %H:%M:%S')
#We rename the column
ozon = ozon.rename(columns={"Endzeit" : "datetime"})
#We set datetime as the index
ozon.set_index("datetime", inplace=True)

In [29]:
#We check our dataset for some information
ozon.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25581 entries, 2021-01-01 01:00:00 to 2023-12-03 00:00:00
Data columns (total 5 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Zürich, Heubeeribüel         25200 non-null  object
 1   Zürich, Rosengartenstrasse   25224 non-null  object
 2   Zürich, Schimmelstrasse      25143 non-null  object
 3   Zürich, Stampfenbachstrasse  25396 non-null  object
 4   Zürich-Kaserne, NABEL        25158 non-null  object
dtypes: object(5)
memory usage: 1.2+ MB


Wir sehen, dass die Daten bis zum 03.12.2023 gehen. Wir müssen also die letzten Tagen droppen (bis zum 30.11.2023 um 23:00). 

In [30]:
#We delete the dates past 2023-11-30 23:00:00
ozon.index = pd.to_datetime(ozon.index)
threshold_date = pd.to_datetime("2023-11-30 23:00:00")
ozon = ozon[ozon.index <= threshold_date]

In [31]:
#We check now our dataset
ozon.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 25532 entries, 2021-01-01 01:00:00 to 2023-11-30 23:00:00
Data columns (total 5 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Zürich, Heubeeribüel         25151 non-null  object
 1   Zürich, Rosengartenstrasse   25175 non-null  object
 2   Zürich, Schimmelstrasse      25094 non-null  object
 3   Zürich, Stampfenbachstrasse  25347 non-null  object
 4   Zürich-Kaserne, NABEL        25109 non-null  object
dtypes: object(5)
memory usage: 1.2+ MB


Wir sollen 25535 Spalten haben. Dies bedeutet, dass irgendwo Daten fehlen. Prüfen wir, was daran falsch ist.

In [32]:
#We adapt the index
ozon.reset_index(level=0, inplace=True)

In [33]:
ozon

datetime Zürich, Heubeeribüel Zürich, Rosengartenstrasse  \
0     2021-01-01 01:00:00                33.11                       4.38   
1     2021-01-01 02:00:00                51.36                       2.25   
2     2021-01-01 03:00:00                 52.2                       4.92   
3     2021-01-01 04:00:00                44.16                      17.73   
4     2021-01-01 05:00:00                46.87                      25.41   
...                   ...                  ...                        ...   
25527 2023-11-30 19:00:00                 17.4                       1.37   
25528 2023-11-30 20:00:00                 6.27                       1.35   
25529 2023-11-30 21:00:00                19.91                       1.28   
25530 2023-11-30 22:00:00                35.81                       1.33   
25531 2023-11-30 23:00:00                29.23                       1.22   

      Zürich, Schimmelstrasse Zürich, Stampfenbachstrasse  \
0                        6.32                        8.51   
1                        6.93                        5.34   
2                        8.83                        7.54   
3                       23.68                        31.3   
4                       24.59                       40.87   
...                       ...                         ...   
25527                    2.17                         2.9   
25528                    2.11                        1.99   
25529                    2.59                        3.28   
25530                    4.59                        6.96   
25531                    1.95                        2.42   

      Zürich-Kaserne, NABEL  
0                      8.39  
1                      5.62  
2                      10.2  
3                      31.2  
4                     33.75  
...                     ...  
25527                  2.74  
25528                  2.96  
25529                  3.74  
25530                 11.56  
25531                   2.3  

[25532 rows x 6 columns]

Um zu erfinden, ob irgendwo Daten fehlen, wurde die folgende Funktion hergestellt. Diese analysiert, ob die Spalte Datetime kontinuierlich ist. Auf diese Weise können wir überprüfen, ob es Probleme im Datensatz gibt.

In [34]:
ozon['datetime'] = pd.to_datetime(ozon['datetime'])

# Check if the 'datetime' column is continuous
is_continuous = (ozon['datetime'] - ozon['datetime'].shift(1)) == pd.Timedelta('1 hour')

# Check if all values in the 'is_continuous' column are True
is_continuous_overall = is_continuous.all()

# Print the result
if is_continuous_overall:
    print("The 'datetime' column is continuous.")
else:
    print("The 'datetime' column is not continuous.")

#This function was created with ChatGPT. This function was also used later in this notebook. This initial citation applies also to those. 
#Question: How can I create a function to check if the datetime column is continuous?

The 'datetime' column is not continuous.


In [35]:
ozon['datetime'] = pd.to_datetime(ozon['datetime'])

# Check if the 'datetime' column is continuous
is_continuous = (ozon['datetime'] - ozon['datetime'].shift(1)) == pd.Timedelta('1 hour')

# Find rows where 'datetime' is not continuous
not_continuous_rows = ozon[~is_continuous]

# Print the rows where 'datetime' is not continuous
print("Rows where 'datetime' is not continuous:")
print(not_continuous_rows)

#This code was generated by ChatGPT. Question: How can find out which rows are problematic?

Rows where 'datetime' is not continuous:
                 datetime Zürich, Heubeeribüel Zürich, Rosengartenstrasse  \
0     2021-01-01 01:00:00                33.11                       4.38   
2066  2021-03-28 04:00:00                49.41                      23.99   
7272  2021-10-31 03:00:00                38.54                       1.16   
7273  2021-10-31 03:00:00                38.54                       1.16   
10801 2022-03-27 04:00:00               104.41                      73.82   
16007 2022-10-30 03:00:00                   28                       1.63   
16008 2022-10-30 03:00:00                   28                       1.63   
19536 2023-03-26 04:00:00                59.71                      35.36   
24742 2023-10-29 03:00:00                60.06                       9.17   
24743 2023-10-29 03:00:00                60.06                       9.17   

      Zürich, Schimmelstrasse Zürich, Stampfenbachstrasse  \
0                        6.32                     

Ich habe mich das ursprüngliche Dataset angeschaut.
- Reihe 2066: es fehlt einen Datenpunkt. Dieses Problem wird durch ein outer merge behoben.
- Reihen 7272 and 7273: es besteht einen Fehler im Dataset. Für die Reihe 7272 muss ich die Zeit anpassen (von 03:00 zu 02:00). 
- Reihe 10801: es fehlt einen Datenpunkt. Dieses Problem wird durch ein outer merge behoben.
- Reihen 16007 and 16008: es besteht einen Fehler im Dataset. Für die Reihe 16007 muss ich die Zeit anpassen (von 03:00 zu 02:00).
- Reihe 19536: es fehlt einen Datenpunkt. Dieses Problem wird durch ein outer merge behoben.
- Reihen 24742 and 24743: es besteht einen Fehler im Dataset. Für die Reihe 24742 muss ich die Zeit anpassen (von 03:00 zu 02:00).

In [36]:
#We need change the time for three rows
ozon.at[7272, 'datetime'] = pd.to_datetime('2021-10-31 02:00:00')
ozon.at[16007, 'datetime'] = pd.to_datetime('2022-10-30 02:00:00')
ozon.at[24742, 'datetime'] = pd.to_datetime('2023-10-29 02:00:00')

In [37]:
ozon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25532 entries, 0 to 25531
Data columns (total 6 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   datetime                     25532 non-null  datetime64[ns]
 1   Zürich, Heubeeribüel         25151 non-null  object        
 2   Zürich, Rosengartenstrasse   25175 non-null  object        
 3   Zürich, Schimmelstrasse      25094 non-null  object        
 4   Zürich, Stampfenbachstrasse  25347 non-null  object        
 5   Zürich-Kaserne, NABEL        25109 non-null  object        
dtypes: datetime64[ns](1), object(5)
memory usage: 1.2+ MB


In [38]:
#We change the index
ozon.reset_index(level=None, drop=False, inplace=True, col_level=0, col_fill="")
#We delete station "Zürich-Kaserne, Nabel", "Zürich, Heubeeribüel", "Zürich, Rosengartenstrasse" and "Zürich, Schimmelstrasse" since we won't use it
ozon = ozon.drop(columns=["Zürich-Kaserne, NABEL", "Zürich, Heubeeribüel", "Zürich, Rosengartenstrasse", "Zürich, Schimmelstrasse"], axis=1)
#We make certain columns to floats to work with them
columns_num_o = ['Zürich, Stampfenbachstrasse']
ozon[columns_num_o] = ozon[columns_num_o].apply(pd.to_numeric, errors='coerce')
#We set datetime as index again
ozon.set_index("datetime", inplace=True)

In [39]:
ozon.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 25532 entries, 2021-01-01 01:00:00 to 2023-11-30 23:00:00
Data columns (total 2 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   index                        25532 non-null  int64  
 1   Zürich, Stampfenbachstrasse  25347 non-null  float64
dtypes: float64(1), int64(1)
memory usage: 598.4 KB


Jetzt ist das Dataset für Ozon bereit und fertig. Die Missing Rows werden wir durch Interpolation und Outer Merge hinzufügen.

### Cleaning Traffic Data

In [40]:
#We load the data
verkehr2021 = pd.read_csv(path + "Verkehr2021.csv", low_memory=False)
verkehr2022 = pd.read_csv(path + "Verkehr2022.csv", low_memory=False)
verkehr2023 = pd.read_csv(path + "Verkehr2023.csv", low_memory=False)

In [41]:
#We add all "verkehr" data to one dataframe
verkehr = pd.concat([verkehr2021, verkehr2022, verkehr2023], ignore_index=True)
#We change the date variable to a useful format
verkehr["MessungDatZeit"] = verkehr["MessungDatZeit"].str.replace('T', ' ')
#We rename the column
verkehr = verkehr.rename(columns={"MessungDatZeit" : "datetime"})
#We set datetime as the index
verkehr.set_index("datetime", inplace=True)

In [42]:
#We take a look at the dataframe
verkehr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4647150 entries, 2021-01-01 00:00:00 to 2023-12-02 23:00:00
Data columns (total 20 columns):
 #   Column              Dtype  
---  ------              -----  
 0   MSID                object 
 1   MSName              object 
 2   ZSID                object 
 3   ZSName              object 
 4   Achse               object 
 5   HNr                 object 
 6   Hoehe               object 
 7   EKoord              float64
 8   NKoord              float64
 9   Richtung            object 
 10  Knummer             int64  
 11  Kname               object 
 12  AnzDetektoren       int64  
 13  D1ID                int64  
 14  D2ID                object 
 15  D3ID                object 
 16  D4ID                object 
 17  LieferDat           object 
 18  AnzFahrzeuge        float64
 19  AnzFahrzeugeStatus  object 
dtypes: float64(3), int64(3), object(14)
memory usage: 744.6+ MB


In [43]:
#We change the index
verkehr.reset_index(level=None, drop=False, inplace=True, col_level=0, col_fill="")
#We delete all objects with no information (only "Unbekannt") in it
verkehr = verkehr.drop(["MSName", "D4ID"], axis=1)
#We delete columns with information that is not needed afterwards for the prediction or merging
verkehr = verkehr.drop(["MSID", "ZSID", "HNr", "Richtung", "Knummer", "Kname", "AnzDetektoren", "D1ID", 
"LieferDat", "AnzFahrzeugeStatus", "Hoehe", "D2ID", "D3ID", "Achse"], axis=1)
#We set datetime as index again
verkehr.set_index("datetime", inplace=True)

In [44]:
#We get the info of the final variables in the dataset
verkehr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4647150 entries, 2021-01-01 00:00:00 to 2023-12-02 23:00:00
Data columns (total 4 columns):
 #   Column        Dtype  
---  ------        -----  
 0   ZSName        object 
 1   EKoord        float64
 2   NKoord        float64
 3   AnzFahrzeuge  float64
dtypes: float64(3), object(1)
memory usage: 177.3+ MB


In [45]:
#We adapt the index to work with the data
verkehr.reset_index(level=None, drop=False, inplace=True, col_level=0, col_fill="")

In [46]:
verkehr["datetime"] = pd.to_datetime(verkehr["datetime"])

#We Filter rows with datetime less than or equal to '2023-11-30 23:00:00'
verkehr = verkehr[verkehr["datetime"] <= "2023-11-30 23:00:00"]

In [47]:
verkehr.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4638318 entries, 0 to 4647101
Data columns (total 5 columns):
 #   Column        Dtype         
---  ------        -----         
 0   datetime      datetime64[ns]
 1   ZSName        object        
 2   EKoord        float64       
 3   NKoord        float64       
 4   AnzFahrzeuge  float64       
dtypes: datetime64[ns](1), float64(3), object(1)
memory usage: 212.3+ MB


Der Datensatz enthält viele Stationen. Für unser Projekt haben wir die Stationen ausgewählt, die der Ozonstation am nächsten liegen. Diese sind: Kornhausbrücke, Milchbucktunnel, Neumühlequai, Stampfenbachstrasse und Walchebrücke.

In [48]:
#We set the desired street
desired_street = "Kornhausbrücke"
#We need to filter the data for the desired street
filtered_data = verkehr[verkehr['ZSName'] == desired_street][['ZSName', 'EKoord', 'NKoord']]
#We print the EKoord and NKoord of the corresponding street name
unique_coordinates = filtered_data.drop_duplicates(subset=['EKoord', 'NKoord'])
unique_coordinates

ZSName      EKoord      NKoord
435550  Kornhausbrücke  2682767.55  1249026.65
444261  Kornhausbrücke  2682755.37  1249023.49

In [49]:
#We select one of the stations
verkehr_sta1 = verkehr[(verkehr['ZSName'] == "Kornhausbrücke") & (verkehr['EKoord'] == 2682767.55) & (verkehr['NKoord'] == 1249026.65)]
verkehr_sta1 = verkehr_sta1.drop(columns=['EKoord', 'NKoord'])

In [50]:
verkehr_sta1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25437 entries, 435550 to 3581968
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   datetime      25437 non-null  datetime64[ns]
 1   ZSName        25437 non-null  object        
 2   AnzFahrzeuge  25396 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 794.9+ KB


In [51]:
#We set the desired street
desired_street = "Milchbucktunnel (Südportal)"
#We need to filter the data for the desired street
filtered_data = verkehr[verkehr['ZSName'] == desired_street][['ZSName', 'EKoord', 'NKoord']]
#We print the EKoord and NKoord of the corresponding street name
unique_coordinates = filtered_data.drop_duplicates(subset=['EKoord', 'NKoord'])
unique_coordinates

ZSName     EKoord      NKoord
522660  Milchbucktunnel (Südportal)  2682974.4  1248893.38
531371  Milchbucktunnel (Südportal)  2682953.4  1248912.91

In [52]:
#We select one of the stations
verkehr_sta2 = verkehr[(verkehr['ZSName'] == "Milchbucktunnel (Südportal)") & (verkehr['EKoord'] == 2682974.4) & (verkehr['NKoord'] == 1248893.38)]
verkehr_sta2 = verkehr_sta2.drop(columns=['EKoord', 'NKoord'])

In [53]:
#We set the desired street
desired_street = "Neumühlequai (Sanität)"
#We need to filter the data for the desired street
filtered_data = verkehr[verkehr['ZSName'] == desired_street][['ZSName', 'EKoord', 'NKoord']]
#We print the EKoord and NKoord of the corresponding street name
unique_coordinates = filtered_data.drop_duplicates(subset=['EKoord', 'NKoord'])
unique_coordinates

ZSName      EKoord      NKoord
958210  Neumühlequai (Sanität)  2683238.97  1248431.30
966921  Neumühlequai (Sanität)  2683200.49  1248492.25

In [54]:
#We select one of the stations
verkehr_sta3 = verkehr[(verkehr['ZSName'] == "Neumühlequai (Sanität)") & (verkehr['EKoord'] == 2683238.97) & (verkehr['NKoord'] == 1248431.30)]
verkehr_sta3 = verkehr_sta3.drop(columns=['EKoord', 'NKoord'])

In [55]:
#We set the desired street
desired_street = "Stampfenbachstrasse (Wasserwerkstrasse)"
#We need to filter the data for the desired street
filtered_data = verkehr[verkehr['ZSName'] == desired_street][['ZSName', 'EKoord', 'NKoord']]
#We print the EKoord and NKoord of the corresponding street name
unique_coordinates = filtered_data.drop_duplicates(subset=['EKoord', 'NKoord'])
unique_coordinates

ZSName      EKoord      NKoord
940788  Stampfenbachstrasse (Wasserwerkstrasse)  2683379.96  1248329.67
949499  Stampfenbachstrasse (Wasserwerkstrasse)  2683370.80  1248325.77

In [56]:
#We select one of the stations
verkehr_sta4 = verkehr[(verkehr['ZSName'] == "Stampfenbachstrasse (Wasserwerkstrasse)") & (verkehr['EKoord'] == 2683379.96) & (verkehr['NKoord'] == 1248329.67)]
verkehr_sta4 = verkehr_sta4.drop(columns=['EKoord', 'NKoord'])

In [57]:
desired_street = "Walchebrücke"

# Filtern Sie die Daten für die spezifizierte Straße
filtered_data = verkehr[verkehr['ZSName'] == desired_street][['ZSName', 'EKoord', 'NKoord']]

# Entfernen Sie doppelte Kombinationen aus 'Ecoord' und 'Ncoord'
unique_coordinates = filtered_data.drop_duplicates(subset=['EKoord', 'NKoord'])
unique_coordinates

ZSName      EKoord      NKoord
452972  Walchebrücke  2683350.45  1248117.80
461683  Walchebrücke  2683297.01  1248118.51

In [58]:
verkehr_sta5 = verkehr[(verkehr['ZSName'] == "Walchebrücke") & (verkehr['EKoord'] == 2683350.45) & (verkehr['NKoord'] == 1248117.80)]
verkehr_sta5 = verkehr_sta5.drop(columns=['EKoord', 'NKoord'])

In [59]:
#We rename the columns to avoid NaN when concatting
verkehr_sta1 = verkehr_sta1.rename(columns={'ZSName': 'ZSName1', 
'Achse': 'Achse1', 'AnzFahrzeuge': 'AnzFahrzeuge1'})
verkehr_sta2 = verkehr_sta2.rename(columns={'ZSName': 'ZSName2', 
'Achse': 'Achse2', 'AnzFahrzeuge': 'AnzFahrzeuge2'})
verkehr_sta3 = verkehr_sta3.rename(columns={'ZSName': 'ZSName3', 
'Achse': 'Achse3', 'AnzFahrzeuge': 'AnzFahrzeuge3'})
verkehr_sta4 = verkehr_sta4.rename(columns={'ZSName': 'ZSName4', 
'Achse': 'Achse4', 'AnzFahrzeuge': 'AnzFahrzeuge4'})
verkehr_sta5 = verkehr_sta5.rename(columns={'ZSName': 'ZSName5', 
'Achse': 'Achse5', 'AnzFahrzeuge': 'AnzFahrzeuge5'})
#We reset the index to avoid Nan when concatting
verkehr_sta1 = verkehr_sta1.reset_index(drop=True)
verkehr_sta2 = verkehr_sta2.reset_index(drop=True)
verkehr_sta3 = verkehr_sta3.reset_index(drop=True)
verkehr_sta4 = verkehr_sta4.reset_index(drop=True)
verkehr_sta5 = verkehr_sta5.reset_index(drop=True)

In [60]:
#We do an outer merge on the datetiem column
dataframes = [verkehr_sta1, verkehr_sta2, verkehr_sta3, verkehr_sta4, verkehr_sta5]

# Perform outer merge on the 'datetime' column
verkehr_sta = dataframes[0]
for df in dataframes[1:]:
    verkehr_sta = pd.merge(verkehr_sta, df, how='outer', on='datetime')
#This code was generated through ChatGPT. Question: How can I do an outer merge with many datasets on the datetime column?

In [61]:
verkehr_sta

datetime         ZSName1  AnzFahrzeuge1  \
0     2021-01-01 00:00:00  Kornhausbrücke          122.0   
1     2021-01-01 01:00:00  Kornhausbrücke          198.0   
2     2021-01-01 02:00:00  Kornhausbrücke          147.0   
3     2021-01-01 03:00:00  Kornhausbrücke          128.0   
4     2021-01-01 04:00:00  Kornhausbrücke          109.0   
...                   ...             ...            ...   
25432 2023-11-30 19:00:00  Kornhausbrücke          358.0   
25433 2023-11-30 20:00:00  Kornhausbrücke          291.0   
25434 2023-11-30 21:00:00  Kornhausbrücke          299.0   
25435 2023-11-30 22:00:00  Kornhausbrücke          275.0   
25436 2023-11-30 23:00:00  Kornhausbrücke          190.0   

                           ZSName2  AnzFahrzeuge2                 ZSName3  \
0      Milchbucktunnel (Südportal)          376.0  Neumühlequai (Sanität)   
1      Milchbucktunnel (Südportal)          421.0  Neumühlequai (Sanität)   
2      Milchbucktunnel (Südportal)          288.0  Neumühlequai (Sanität)   
3      Milchbucktunnel (Südportal)          188.0  Neumühlequai (Sanität)   
4      Milchbucktunnel (Südportal)          133.0  Neumühlequai (Sanität)   
...                            ...            ...                     ...   
25432  Milchbucktunnel (Südportal)          938.0  Neumühlequai (Sanität)   
25433  Milchbucktunnel (Südportal)          824.0  Neumühlequai (Sanität)   
25434  Milchbucktunnel (Südportal)          617.0  Neumühlequai (Sanität)   
25435  Milchbucktunnel (Südportal)          583.0  Neumühlequai (Sanität)   
25436  Milchbucktunnel (Südportal)          369.0  Neumühlequai (Sanität)   

       AnzFahrzeuge3                                  ZSName4  AnzFahrzeuge4  \
0              252.0  Stampfenbachstrasse (Wasserwerkstrasse)          167.0   
1              240.0  Stampfenbachstrasse (Wasserwerkstrasse)          220.0   
2              171.0  Stampfenbachstrasse (Wasserwerkstrasse)          142.0   
3              113.0  Stampfenbachstrasse (Wasserwerkstrasse)           84.0   
4               68.0  Stampfenbachstrasse (Wasserwerkstrasse)           65.0   
...              ...                                      ...            ...   
25432          474.0  Stampfenbachstrasse (Wasserwerkstrasse)          416.0   
25433          435.0  Stampfenbachstrasse (Wasserwerkstrasse)          332.0   
25434          351.0  Stampfenbachstrasse (Wasserwerkstrasse)          284.0   
25435          336.0  Stampfenbachstrasse (Wasserwerkstrasse)          252.0   
25436          219.0  Stampfenbachstrasse (Wasserwerkstrasse)          199.0   

            ZSName5  AnzFahrzeuge5  
0      Walchebrücke          232.0  
1      Walchebrücke          241.0  
2      Walchebrücke          171.0  
3      Walchebrücke          104.0  
4      Walchebrücke           70.0  
...             ...            ...  
25432  Walchebrücke          556.0  
25433  Walchebrücke          475.0  
25434  Walchebrücke          395.0  
25435  Walchebrücke          362.0  
25436  Walchebrücke          263.0  

[25437 rows x 11 columns]

In [62]:
#We check if the dataset is continuous
verkehr_sta['datetime'] = pd.to_datetime(verkehr_sta['datetime'])

# Check if the 'datetime' column is continuous
is_continuous = (verkehr_sta['datetime'] - verkehr_sta['datetime'].shift(1)) == pd.Timedelta('1 hour')

# Check if all values in the 'is_continuous' column are True
is_continuous_overall = is_continuous.all()

# Print the result
if is_continuous_overall:
    print("The 'datetime' column is continuous.")
else:
    print("The 'datetime' column is not continuous.")

The 'datetime' column is not continuous.


In [63]:
#We check where the dataset is not continuous
verkehr_sta['datetime'] = pd.to_datetime(verkehr_sta['datetime'])

# Check if the 'datetime' column is continuous
is_continuous = (verkehr_sta['datetime'] - verkehr_sta['datetime'].shift(1)) == pd.Timedelta('1 hour')

# Find rows where 'datetime' is not continuous
not_continuous_rows = verkehr_sta[~is_continuous]

# Print the rows where 'datetime' is not continuous
print("Rows where 'datetime' is not continuous:")
print(not_continuous_rows)

Rows where 'datetime' is not continuous:
                 datetime         ZSName1  AnzFahrzeuge1  \
0     2021-01-01 00:00:00  Kornhausbrücke          122.0   
2066  2021-03-28 03:00:00  Kornhausbrücke          115.0   
8711  2022-01-01 00:00:00  Kornhausbrücke          202.0   
10753 2022-03-27 03:00:00  Kornhausbrücke          327.0   
17422 2023-01-01 00:00:00  Kornhausbrücke          247.0   
19440 2023-03-26 03:00:00  Kornhausbrücke          268.0   

                           ZSName2  AnzFahrzeuge2                 ZSName3  \
0      Milchbucktunnel (Südportal)          376.0  Neumühlequai (Sanität)   
2066   Milchbucktunnel (Südportal)          242.0  Neumühlequai (Sanität)   
8711   Milchbucktunnel (Südportal)          767.0  Neumühlequai (Sanität)   
10753  Milchbucktunnel (Südportal)          583.0  Neumühlequai (Sanität)   
17422  Milchbucktunnel (Südportal)          685.0  Neumühlequai (Sanität)   
19440  Milchbucktunnel (Südportal)          518.0  Neumühlequai (Sanität)   

Es bestehen die folgenden Probleme:
- Reihe 2066: Es fehlt den Datenpunkt 2021-03-28 02:00:00
- Reihe 8711: Es fehlen Daten bis zum 2021-12-31 23:00:00
- Reihe 10753: Es fehlt den Datenpunkt 2022-03-27 02:00:00
- Reihe 17422: Es fehlen Daten bis zum 2022-12-31 23:00:00
- Reihe 19440: Es fehlt den Datenpunkt 2023-03-26 02:00:00

In [64]:
verkehr_sta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25437 entries, 0 to 25436
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   datetime       25437 non-null  datetime64[ns]
 1   ZSName1        25437 non-null  object        
 2   AnzFahrzeuge1  25396 non-null  float64       
 3   ZSName2        25437 non-null  object        
 4   AnzFahrzeuge2  25262 non-null  float64       
 5   ZSName3        25437 non-null  object        
 6   AnzFahrzeuge3  25406 non-null  float64       
 7   ZSName4        25437 non-null  object        
 8   AnzFahrzeuge4  25390 non-null  float64       
 9   ZSName5        25437 non-null  object        
 10  AnzFahrzeuge5  25413 non-null  float64       
dtypes: datetime64[ns](1), float64(5), object(5)
memory usage: 2.1+ MB


In [65]:
#We drop the unnecessary columns
verkehr_sta = verkehr_sta.drop(columns= ["ZSName1", "ZSName2", "ZSName3", "ZSName4", "ZSName5"])

In [66]:
#We do interpolation
all_variables_vsta = (list(verkehr_sta.columns))
verkehr_sta = verkehr_sta[all_variables_vsta].interpolate(option='spline')

In [67]:
verkehr_sta

datetime  AnzFahrzeuge1  AnzFahrzeuge2  AnzFahrzeuge3  \
0     2021-01-01 00:00:00          122.0          376.0          252.0   
1     2021-01-01 01:00:00          198.0          421.0          240.0   
2     2021-01-01 02:00:00          147.0          288.0          171.0   
3     2021-01-01 03:00:00          128.0          188.0          113.0   
4     2021-01-01 04:00:00          109.0          133.0           68.0   
...                   ...            ...            ...            ...   
25432 2023-11-30 19:00:00          358.0          938.0          474.0   
25433 2023-11-30 20:00:00          291.0          824.0          435.0   
25434 2023-11-30 21:00:00          299.0          617.0          351.0   
25435 2023-11-30 22:00:00          275.0          583.0          336.0   
25436 2023-11-30 23:00:00          190.0          369.0          219.0   

       AnzFahrzeuge4  AnzFahrzeuge5  
0              167.0          232.0  
1              220.0          241.0  
2              142.0          171.0  
3               84.0          104.0  
4               65.0           70.0  
...              ...            ...  
25432          416.0          556.0  
25433          332.0          475.0  
25434          284.0          395.0  
25435          252.0          362.0  
25436          199.0          263.0  

[25437 rows x 6 columns]

In [68]:
#We sum up the count of vehicles
verkehr_sta['AnzFahrzeuge_STA'] = verkehr_sta[['AnzFahrzeuge1', 'AnzFahrzeuge2', 'AnzFahrzeuge3', 'AnzFahrzeuge4', 'AnzFahrzeuge5']].sum(axis=1)
#We remove the columns not needed anymore
verkehr_sta = verkehr_sta[["datetime", "AnzFahrzeuge_STA"]]
# rename the datetime column
verkehr_sta = verkehr_sta.rename(columns={'datetime': 'datetime_sta'})
# convert float into integer while only counting whole vehicles
verkehr_sta['AnzFahrzeuge_STA'] = verkehr_sta['AnzFahrzeuge_STA'].astype(int)

In [69]:
verkehr_sta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25437 entries, 0 to 25436
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   datetime_sta      25437 non-null  datetime64[ns]
 1   AnzFahrzeuge_STA  25437 non-null  int32         
dtypes: datetime64[ns](1), int32(1)
memory usage: 298.2 KB


In [70]:
#we drop the first row, as we consider data from 01:00:00
verkehr_sta = verkehr_sta.iloc[1:]

In [71]:
verkehr_sta['datetime_sta'] = pd.to_datetime(verkehr_sta['datetime_sta'])

#We check if the 'datetime' column is continuous
is_continuous = (verkehr_sta['datetime_sta'] - verkehr_sta['datetime_sta'].shift(1)) == pd.Timedelta('1 hour')

#We check if all values in the 'is_continuous' column are True
is_continuous_overall = is_continuous.all()

#We print the result
if is_continuous_overall:
    print("The 'datetime' column is continuous.")
else:
    print("The 'datetime' column is not continuous.")

The 'datetime' column is not continuous.


In [72]:
verkehr_sta['datetime_sta'] = pd.to_datetime(verkehr_sta['datetime_sta'])

# Check if the 'datetime' column is continuous
is_continuous = (verkehr_sta['datetime_sta'] - verkehr_sta['datetime_sta'].shift(1)) == pd.Timedelta('1 hour')

# Find rows where 'datetime' is not continuous
not_continuous_rows = verkehr_sta[~is_continuous]

# Print the rows where 'datetime' is not continuous
print("Rows where 'datetime' is not continuous:")
print(not_continuous_rows)

Rows where 'datetime' is not continuous:
             datetime_sta  AnzFahrzeuge_STA
1     2021-01-01 01:00:00              1320
2066  2021-03-28 03:00:00               718
8711  2022-01-01 00:00:00              2377
10753 2022-03-27 03:00:00              1855
17422 2023-01-01 00:00:00              2218
19440 2023-03-26 03:00:00              1713


# Merging of Ozon, Weather and Verkehr Data

In [73]:
#We reset the index
ozon.reset_index(level=None, drop=False, inplace=True, col_level=0, col_fill="")

In [74]:
ozon

datetime  index  Zürich, Stampfenbachstrasse
0     2021-01-01 01:00:00      0                         8.51
1     2021-01-01 02:00:00      1                         5.34
2     2021-01-01 03:00:00      2                         7.54
3     2021-01-01 04:00:00      3                        31.30
4     2021-01-01 05:00:00      4                        40.87
...                   ...    ...                          ...
25527 2023-11-30 19:00:00  25527                         2.90
25528 2023-11-30 20:00:00  25528                         1.99
25529 2023-11-30 21:00:00  25529                         3.28
25530 2023-11-30 22:00:00  25530                         6.96
25531 2023-11-30 23:00:00  25531                         2.42

[25532 rows x 3 columns]

In [75]:
#We reset the index
wetter.reset_index(level=None, drop=False, inplace=True, col_level=0, col_fill="")

In [76]:
wetter

51074             datetime  Luftdruck_STA  Lufttemp_STA  Windgeschw_STA
0      2021-01-01 01:00:00          954.0           2.1             4.7
1      2021-01-01 02:00:00          954.4           2.3             3.2
2      2021-01-01 03:00:00          954.6           2.0             2.5
3      2021-01-01 04:00:00          954.6           2.3             2.2
4      2021-01-01 05:00:00          954.8           2.3             2.9
...                    ...            ...           ...             ...
25530  2023-11-30 19:00:00          948.9           3.2             2.9
25531  2023-11-30 20:00:00          948.9           3.8             5.4
25532  2023-11-30 21:00:00          949.1           4.3             6.8
25533  2023-11-30 22:00:00          949.2           3.8             5.4
25534  2023-11-30 23:00:00          949.1           4.6             4.7

[25535 rows x 4 columns]

In [77]:
#We convert the datetime column to_datetime
wetter['datetime'] = pd.to_datetime(wetter['datetime'])
ozon['datetime'] = pd.to_datetime(ozon['datetime'])

#We do an outer merge between the wetter and the ozon dataset
df_merged = pd.merge(wetter, ozon, on='datetime', how='outer')

In [78]:
df_merged

datetime  Luftdruck_STA  Lufttemp_STA  Windgeschw_STA  \
0     2021-01-01 01:00:00          954.0           2.1             4.7   
1     2021-01-01 02:00:00          954.4           2.3             3.2   
2     2021-01-01 03:00:00          954.6           2.0             2.5   
3     2021-01-01 04:00:00          954.6           2.3             2.2   
4     2021-01-01 05:00:00          954.8           2.3             2.9   
...                   ...            ...           ...             ...   
25530 2023-11-30 19:00:00          948.9           3.2             2.9   
25531 2023-11-30 20:00:00          948.9           3.8             5.4   
25532 2023-11-30 21:00:00          949.1           4.3             6.8   
25533 2023-11-30 22:00:00          949.2           3.8             5.4   
25534 2023-11-30 23:00:00          949.1           4.6             4.7   

         index  Zürich, Stampfenbachstrasse  
0          0.0                         8.51  
1          1.0                         5.34  
2          2.0                         7.54  
3          3.0                        31.30  
4          4.0                        40.87  
...        ...                          ...  
25530  25527.0                         2.90  
25531  25528.0                         1.99  
25532  25529.0                         3.28  
25533  25530.0                         6.96  
25534  25531.0                         2.42  

[25535 rows x 6 columns]

Jetzt haben wir die richtige Anzahl von Spalten.

In [79]:
#We rename the datetime column of the verkehr dataset
verkehr_sta = verkehr_sta.rename(columns={'datetime_sta': 'datetime'})

In [80]:
verkehr_sta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25436 entries, 1 to 25436
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   datetime          25436 non-null  datetime64[ns]
 1   AnzFahrzeuge_STA  25436 non-null  int32         
dtypes: datetime64[ns](1), int32(1)
memory usage: 298.2 KB


In [81]:
#We change the datetime column to_datetime
verkehr_sta['datetime'] = pd.to_datetime(verkehr_sta['datetime'])

In [82]:
#We do an outer merge between the wetter and ozon dataset and the verkehr dataset
df_complete = pd.merge(df_merged, verkehr_sta, on='datetime', how='outer')

In [83]:
#We check our merged dataset
df_complete.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25535 entries, 0 to 25534
Data columns (total 7 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   datetime                     25535 non-null  datetime64[ns]
 1   Luftdruck_STA                25535 non-null  float64       
 2   Lufttemp_STA                 25535 non-null  float64       
 3   Windgeschw_STA               25535 non-null  float64       
 4   index                        25532 non-null  float64       
 5   Zürich, Stampfenbachstrasse  25347 non-null  float64       
 6   AnzFahrzeuge_STA             25436 non-null  float64       
dtypes: datetime64[ns](1), float64(6)
memory usage: 1.4 MB


In [84]:
#We do interpolation
all_variables = (list(df_complete.columns))
df_complete = df_complete[all_variables].interpolate(option='spline')

In [85]:
#We get the information for the merged dataset
df_complete.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25535 entries, 0 to 25534
Data columns (total 7 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   datetime                     25535 non-null  datetime64[ns]
 1   Luftdruck_STA                25535 non-null  float64       
 2   Lufttemp_STA                 25535 non-null  float64       
 3   Windgeschw_STA               25535 non-null  float64       
 4   index                        25535 non-null  float64       
 5   Zürich, Stampfenbachstrasse  25535 non-null  float64       
 6   AnzFahrzeuge_STA             25535 non-null  float64       
dtypes: datetime64[ns](1), float64(6)
memory usage: 1.4 MB


In [86]:
#We add some time columns for forecasting
#We convert the 'datetime_column' to a datetime object
df_complete['datetime'] = pd.to_datetime(df_complete['datetime'], format='%Y-%m-%d %H:%M:%S')
#We extract individual components
df_complete['hour_of_day'] = df_complete['datetime'].dt.hour
df_complete['day_of_week'] = df_complete['datetime'].dt.dayofweek + 1  #Monday is 0, so add 1
df_complete['day_of_year'] = df_complete['datetime'].dt.dayofyear
df_complete['month'] = df_complete['datetime'].dt.month
df_complete['year'] = df_complete['datetime'].dt.year
#We reset the index
df_complete.set_index("datetime", inplace=True)

In [87]:
df_complete.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 25535 entries, 2021-01-01 01:00:00 to 2023-11-30 23:00:00
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Luftdruck_STA                25535 non-null  float64
 1   Lufttemp_STA                 25535 non-null  float64
 2   Windgeschw_STA               25535 non-null  float64
 3   index                        25535 non-null  float64
 4   Zürich, Stampfenbachstrasse  25535 non-null  float64
 5   AnzFahrzeuge_STA             25535 non-null  float64
 6   hour_of_day                  25535 non-null  int32  
 7   day_of_week                  25535 non-null  int32  
 8   day_of_year                  25535 non-null  int32  
 9   month                        25535 non-null  int32  
 10  year                         25535 non-null  int32  
dtypes: float64(6), int32(5)
memory usage: 1.9 MB


In [88]:
#We change the name of the ozon column
df_complete = df_complete.rename(columns={'Zürich, Stampfenbachstrasse': 'StampfenbachstrOzon'})

### Creating the Dataframe

In [89]:
#We create a dataframe for "Stampfenbachstrasse"
#We select the right data
df_sta = df_complete[["Luftdruck_STA", "Lufttemp_STA", "Windgeschw_STA", 
"StampfenbachstrOzon", "AnzFahrzeuge_STA", "hour_of_day", "day_of_week", "day_of_year", "month", "year"]]

In [90]:
#We get the information
df_sta.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 25535 entries, 2021-01-01 01:00:00 to 2023-11-30 23:00:00
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Luftdruck_STA        25535 non-null  float64
 1   Lufttemp_STA         25535 non-null  float64
 2   Windgeschw_STA       25535 non-null  float64
 3   StampfenbachstrOzon  25535 non-null  float64
 4   AnzFahrzeuge_STA     25535 non-null  float64
 5   hour_of_day          25535 non-null  int32  
 6   day_of_week          25535 non-null  int32  
 7   day_of_year          25535 non-null  int32  
 8   month                25535 non-null  int32  
 9   year                 25535 non-null  int32  
dtypes: float64(5), int32(5)
memory usage: 1.7 MB


In [91]:
#We save the dataset
df_sta.to_csv(path + "sta_data.csv", index=True)

In [92]:
df_sta

Luftdruck_STA  Lufttemp_STA  Windgeschw_STA  \
datetime                                                           
2021-01-01 01:00:00          954.0           2.1             4.7   
2021-01-01 02:00:00          954.4           2.3             3.2   
2021-01-01 03:00:00          954.6           2.0             2.5   
2021-01-01 04:00:00          954.6           2.3             2.2   
2021-01-01 05:00:00          954.8           2.3             2.9   
...                            ...           ...             ...   
2023-11-30 19:00:00          948.9           3.2             2.9   
2023-11-30 20:00:00          948.9           3.8             5.4   
2023-11-30 21:00:00          949.1           4.3             6.8   
2023-11-30 22:00:00          949.2           3.8             5.4   
2023-11-30 23:00:00          949.1           4.6             4.7   

                     StampfenbachstrOzon  AnzFahrzeuge_STA  hour_of_day  \
datetime                                                                  
2021-01-01 01:00:00                 8.51            1320.0            1   
2021-01-01 02:00:00                 5.34             919.0            2   
2021-01-01 03:00:00                 7.54             617.0            3   
2021-01-01 04:00:00                31.30             445.0            4   
2021-01-01 05:00:00                40.87             371.0            5   
...                                  ...               ...          ...   
2023-11-30 19:00:00                 2.90            2742.0           19   
2023-11-30 20:00:00                 1.99            2357.0           20   
2023-11-30 21:00:00                 3.28            1946.0           21   
2023-11-30 22:00:00                 6.96            1808.0           22   
2023-11-30 23:00:00                 2.42            1240.0           23   

                     day_of_week  day_of_year  month  year  
datetime                                                    
2021-01-01 01:00:00            5            1      1  2021  
2021-01-01 02:00:00            5            1      1  2021  
2021-01-01 03:00:00            5            1      1  2021  
2021-01-01 04:00:00            5            1      1  2021  
2021-01-01 05:00:00            5            1      1  2021  
...                          ...          ...    ...   ...  
2023-11-30 19:00:00            4          334     11  2023  
2023-11-30 20:00:00            4          334     11  2023  
2023-11-30 21:00:00            4          334     11  2023  
2023-11-30 22:00:00            4          334     11  2023  
2023-11-30 23:00:00            4          334     11  2023  

[25535 rows x 10 columns]

# Forecasting dataset

Da wir Ozon Werte in die Zukunft vorhersagen möchten, müssen wir unser Forecasting Dataset vorbereiten. Das unterstehende Code basiert auf dem folgenden Tutorial: https://www.youtube.com/watch?v=z3ZnOW-S550&t=922s

In [93]:
#We load the necessary packages
from datetime import datetime, timedelta

In [94]:
#We create a function to make lags for the other features
def add_lags_features(df, columns):
    for column_name in columns:
        target_map = df[column_name].to_dict()
        df[column_name] = (df.index - pd.Timedelta("48 hours")).map(target_map)
    return df

In [95]:
#We have a function to add the time series components
def create_features(df):
    """
    Create time series features based on time series index.
    """
    df = df.copy()
    df['hour_of_day'] = df.index.hour
    df['day_of_week'] = df.index.dayofweek + 1
    df['day_of_year'] = df.index.dayofyear
    df['month'] = df.index.month
    df['year'] = df.index.year
    
    
    return df

In [96]:
#We create an empty dataset
future = pd.date_range('2023-12-01', '2023-12-02', freq='1h')
future_df_sta = pd.DataFrame(index=future)
future_df_sta['isFuture'] = True #isFuture will appear for the forecasting part
df_sta['isFuture'] = False
df_sta_and_future = pd.concat([df_sta, future_df_sta]) #We concatenate the old and the new dataset

In [97]:
df_sta_and_future

Luftdruck_STA  Lufttemp_STA  Windgeschw_STA  \
2021-01-01 01:00:00          954.0           2.1             4.7   
2021-01-01 02:00:00          954.4           2.3             3.2   
2021-01-01 03:00:00          954.6           2.0             2.5   
2021-01-01 04:00:00          954.6           2.3             2.2   
2021-01-01 05:00:00          954.8           2.3             2.9   
...                            ...           ...             ...   
2023-12-01 20:00:00            NaN           NaN             NaN   
2023-12-01 21:00:00            NaN           NaN             NaN   
2023-12-01 22:00:00            NaN           NaN             NaN   
2023-12-01 23:00:00            NaN           NaN             NaN   
2023-12-02 00:00:00            NaN           NaN             NaN   

                     StampfenbachstrOzon  AnzFahrzeuge_STA  hour_of_day  \
2021-01-01 01:00:00                 8.51            1320.0          1.0   
2021-01-01 02:00:00                 5.34             919.0          2.0   
2021-01-01 03:00:00                 7.54             617.0          3.0   
2021-01-01 04:00:00                31.30             445.0          4.0   
2021-01-01 05:00:00                40.87             371.0          5.0   
...                                  ...               ...          ...   
2023-12-01 20:00:00                  NaN               NaN          NaN   
2023-12-01 21:00:00                  NaN               NaN          NaN   
2023-12-01 22:00:00                  NaN               NaN          NaN   
2023-12-01 23:00:00                  NaN               NaN          NaN   
2023-12-02 00:00:00                  NaN               NaN          NaN   

                     day_of_week  day_of_year  month    year  isFuture  
2021-01-01 01:00:00          5.0          1.0    1.0  2021.0     False  
2021-01-01 02:00:00          5.0          1.0    1.0  2021.0     False  
2021-01-01 03:00:00          5.0          1.0    1.0  2021.0     False  
2021-01-01 04:00:00          5.0          1.0    1.0  2021.0     False  
2021-01-01 05:00:00          5.0          1.0    1.0  2021.0     False  
...                          ...          ...    ...     ...       ...  
2023-12-01 20:00:00          NaN          NaN    NaN     NaN      True  
2023-12-01 21:00:00          NaN          NaN    NaN     NaN      True  
2023-12-01 22:00:00          NaN          NaN    NaN     NaN      True  
2023-12-01 23:00:00          NaN          NaN    NaN     NaN      True  
2023-12-02 00:00:00          NaN          NaN    NaN     NaN      True  

[25560 rows x 11 columns]

In [98]:
#We apply our functions to create the lags and the time series features
df_sta_and_future = create_features(df_sta_and_future)
columns_to_lag = ["Luftdruck_STA", "Lufttemp_STA", "Windgeschw_STA", "AnzFahrzeuge_STA"]
add_lags_features(df_sta_and_future, columns_to_lag)

Luftdruck_STA  Lufttemp_STA  Windgeschw_STA  \
2021-01-01 01:00:00            NaN           NaN             NaN   
2021-01-01 02:00:00            NaN           NaN             NaN   
2021-01-01 03:00:00            NaN           NaN             NaN   
2021-01-01 04:00:00            NaN           NaN             NaN   
2021-01-01 05:00:00            NaN           NaN             NaN   
...                            ...           ...             ...   
2023-12-01 20:00:00          955.0           3.2             7.2   
2023-12-01 21:00:00          954.5           3.0             6.8   
2023-12-01 22:00:00          954.0           2.9             5.4   
2023-12-01 23:00:00          953.8           2.6             5.8   
2023-12-02 00:00:00          953.6           2.4             3.2   

                     StampfenbachstrOzon  AnzFahrzeuge_STA  hour_of_day  \
2021-01-01 01:00:00                 8.51               NaN            1   
2021-01-01 02:00:00                 5.34               NaN            2   
2021-01-01 03:00:00                 7.54               NaN            3   
2021-01-01 04:00:00                31.30               NaN            4   
2021-01-01 05:00:00                40.87               NaN            5   
...                                  ...               ...          ...   
2023-12-01 20:00:00                  NaN            2477.0           20   
2023-12-01 21:00:00                  NaN            2115.0           21   
2023-12-01 22:00:00                  NaN            1865.0           22   
2023-12-01 23:00:00                  NaN            1050.0           23   
2023-12-02 00:00:00                  NaN             604.0            0   

                     day_of_week  day_of_year  month  year  isFuture  
2021-01-01 01:00:00            5            1      1  2021     False  
2021-01-01 02:00:00            5            1      1  2021     False  
2021-01-01 03:00:00            5            1      1  2021     False  
2021-01-01 04:00:00            5            1      1  2021     False  
2021-01-01 05:00:00            5            1      1  2021     False  
...                          ...          ...    ...   ...       ...  
2023-12-01 20:00:00            5          335     12  2023      True  
2023-12-01 21:00:00            5          335     12  2023      True  
2023-12-01 22:00:00            5          335     12  2023      True  
2023-12-01 23:00:00            5          335     12  2023      True  
2023-12-02 00:00:00            6          336     12  2023      True  

[25560 rows x 11 columns]

In [99]:
#We select our forecasting dataset
df_sta_forecasting = df_sta_and_future[df_sta_and_future['isFuture'] == True]

#We drop the last row
df_sta_forecasting = df_sta_forecasting.iloc[:-1]
df_sta_forecasting = df_sta_forecasting.drop(columns=['isFuture'])
df_sta_forecasting

Luftdruck_STA  Lufttemp_STA  Windgeschw_STA  \
2023-12-01 00:00:00          958.5           1.7             4.0   
2023-12-01 01:00:00          958.3           1.3             3.6   
2023-12-01 02:00:00          958.5           0.8             2.2   
2023-12-01 03:00:00          958.5           0.8             3.6   
2023-12-01 04:00:00          958.1           0.7             3.6   
2023-12-01 05:00:00          958.1           1.0             3.2   
2023-12-01 06:00:00          958.1           1.1             2.9   
2023-12-01 07:00:00          958.4           1.2             4.3   
2023-12-01 08:00:00          958.6           1.5             5.4   
2023-12-01 09:00:00          958.8           1.9             7.6   
2023-12-01 10:00:00          958.8           2.2             7.6   
2023-12-01 11:00:00          958.4           2.5            10.8   
2023-12-01 12:00:00          957.6           3.4             9.0   
2023-12-01 13:00:00          956.8           4.1             5.8   
2023-12-01 14:00:00          956.1           4.4             6.1   
2023-12-01 15:00:00          955.7           4.3             6.1   
2023-12-01 16:00:00          955.5           4.0             5.8   
2023-12-01 17:00:00          955.3           3.6             5.0   
2023-12-01 18:00:00          955.2           3.3             4.7   
2023-12-01 19:00:00          955.2           3.2             5.8   
2023-12-01 20:00:00          955.0           3.2             7.2   
2023-12-01 21:00:00          954.5           3.0             6.8   
2023-12-01 22:00:00          954.0           2.9             5.4   
2023-12-01 23:00:00          953.8           2.6             5.8   

                     StampfenbachstrOzon  AnzFahrzeuge_STA  hour_of_day  \
2023-12-01 00:00:00                  NaN             477.0            0   
2023-12-01 01:00:00                  NaN             365.0            1   
2023-12-01 02:00:00                  NaN             245.0            2   
2023-12-01 03:00:00                  NaN             152.0            3   
2023-12-01 04:00:00                  NaN             255.0            4   
2023-12-01 05:00:00                  NaN             748.0            5   
2023-12-01 06:00:00                  NaN            2307.0            6   
2023-12-01 07:00:00                  NaN            2879.0            7   
2023-12-01 08:00:00                  NaN            2736.0            8   
2023-12-01 09:00:00                  NaN            2523.0            9   
2023-12-01 10:00:00                  NaN            2823.0           10   
2023-12-01 11:00:00                  NaN            3215.0           11   
2023-12-01 12:00:00                  NaN            3136.0           12   
2023-12-01 13:00:00                  NaN            3214.0           13   
2023-12-01 14:00:00                  NaN            3455.0           14   
2023-12-01 15:00:00                  NaN            3768.0           15   
2023-12-01 16:00:00                  NaN            4072.0           16   
2023-12-01 17:00:00                  NaN            4119.0           17   
2023-12-01 18:00:00                  NaN            3824.0           18   
2023-12-01 19:00:00                  NaN            3215.0           19   
2023-12-01 20:00:00                  NaN            2477.0           20   
2023-12-01 21:00:00                  NaN            2115.0           21   
2023-12-01 22:00:00                  NaN            1865.0           22   
2023-12-01 23:00:00                  NaN            1050.0           23   

                     day_of_week  day_of_year  month  year  
2023-12-01 00:00:00            5          335     12  2023  
2023-12-01 01:00:00            5          335     12  2023  
2023-12-01 02:00:00            5          335     12  2023  
2023-12-01 03:00:00            5          335     12  2023  
2023-12-01 04:00:00            5          335     12  2023  
2023-12-01 05:00:00            5          335     12  2023  
2023-12-01 06:00:

In [100]:
#We now concatenate the old and the forecasting dataset
result_df = pd.concat([df_sta, df_sta_forecasting])
#We drop the isFuture column
result_df = result_df.drop(columns=['isFuture'])

In [101]:
result_df

Luftdruck_STA  Lufttemp_STA  Windgeschw_STA  \
2021-01-01 01:00:00          954.0           2.1             4.7   
2021-01-01 02:00:00          954.4           2.3             3.2   
2021-01-01 03:00:00          954.6           2.0             2.5   
2021-01-01 04:00:00          954.6           2.3             2.2   
2021-01-01 05:00:00          954.8           2.3             2.9   
...                            ...           ...             ...   
2023-12-01 19:00:00          955.2           3.2             5.8   
2023-12-01 20:00:00          955.0           3.2             7.2   
2023-12-01 21:00:00          954.5           3.0             6.8   
2023-12-01 22:00:00          954.0           2.9             5.4   
2023-12-01 23:00:00          953.8           2.6             5.8   

                     StampfenbachstrOzon  AnzFahrzeuge_STA  hour_of_day  \
2021-01-01 01:00:00                 8.51            1320.0            1   
2021-01-01 02:00:00                 5.34             919.0            2   
2021-01-01 03:00:00                 7.54             617.0            3   
2021-01-01 04:00:00                31.30             445.0            4   
2021-01-01 05:00:00                40.87             371.0            5   
...                                  ...               ...          ...   
2023-12-01 19:00:00                  NaN            3215.0           19   
2023-12-01 20:00:00                  NaN            2477.0           20   
2023-12-01 21:00:00                  NaN            2115.0           21   
2023-12-01 22:00:00                  NaN            1865.0           22   
2023-12-01 23:00:00                  NaN            1050.0           23   

                     day_of_week  day_of_year  month  year  
2021-01-01 01:00:00            5            1      1  2021  
2021-01-01 02:00:00            5            1      1  2021  
2021-01-01 03:00:00            5            1      1  2021  
2021-01-01 04:00:00            5            1      1  2021  
2021-01-01 05:00:00            5            1      1  2021  
...                          ...          ...    ...   ...  
2023-12-01 19:00:00            5          335     12  2023  
2023-12-01 20:00:00            5          335     12  2023  
2023-12-01 21:00:00            5          335     12  2023  
2023-12-01 22:00:00            5          335     12  2023  
2023-12-01 23:00:00            5          335     12  2023  

[25559 rows x 10 columns]

In [102]:
#We save the dataset
result_df.to_csv(path + "sta_data_forecasting.csv")